beacon-chain/sync/backfill/worker.go

```diff
@@ -9,58 +9,154 @@ import (
+				b = w.handleColumns(ctx, b)
+			case batchSequenced:
+				b = w.handleBlocks(ctx, b)
+			case batchImportable:
+				b.blocks = nil
+				b = w.handleBlocks(ctx, b)
+			default:
+				log.WithFields(b.logFields()).WithField("backfillWorker", w.id).Debug("batch in unhandled state")
+				panic("unhandled batch state") // lint:nopanic -- TODO: this panic is temporary / for debugging.
+			}
+			w.done <- b
 		case <-ctx.Done():
 			log.WithField("backfillWorker", w.id).Info("Backfill worker exiting after context canceled")
 			return
 		}
 	}
 }
 
+func resetRetryableColumns(b batch) batch {
+	// return the given batch as-is if it isn't in a state that this func should handle.
+	if b.columns == nil || b.columns.bisector == nil || len(b.columns.bisector.errs) == 0 {
+		return b
+	}
+	bisector := b.columns.bisector
+	roots := bisector.failingRoots()
+	if len(roots) == 0 {
+		return b
+	}
+	// Add all the failed columns back to the toDownload structure.
+	for _, root := range roots {
+		bc := b.columns.toDownload[root]
+		bc.remaining.Union(bisector.failuresFor(root))
+	}
+	b.columns.bisector.reset()
+	return b
+}
+
 func (w *p2pWorker) handleBlocks(ctx context.Context, b batch) batch {
-	cs := w.c.CurrentSlot()
-	blobRetentionStart, err := sync.BlobRPCMinValidSlot(cs)
+	current := w.cfg.clock.CurrentSlot()
+	// TODO: refactor all the blob and column setup stuff.
+	// we know the slot when we first set up the batch, so we should be able to determine if we need the blob setup bits at all
+	// before we fetch the blocks. Same goes for the column dependencies.
+	blobRetentionStart, err := sync.BlobRPCMinValidSlot(current)
 	if err != nil {
 		return b.withRetryableError(errors.Wrap(err, "configuration issue, could not compute minimum blob retention slot"))
 	}
-	b.blockPid = b.busy
+	b.blockPid = b.peer
 	start := time.Now()
-	results, err := sync.SendBeaconBlocksByRangeRequest(ctx, w.c, w.p2p, b.blockPid, b.blockRequest(), blockValidationMetrics)
-	dlt := time.Now()
-	backfillBatchTimeDownloadingBlocks.Observe(float64(dlt.Sub(start).Milliseconds()))
+	results, err := sync.SendBeaconBlocksByRangeRequest(ctx, w.cfg.clock, w.p2p, b.blockPid, b.blockRequest(), blockValidationMetrics)
 	if err != nil {
 		log.WithError(err).WithFields(b.logFields()).Debug("Batch requesting failed")
 		return b.withRetryableError(err)
 	}
-	vb, err := w.v.verify(results)
-	backfillBatchTimeVerifying.Observe(float64(time.Since(dlt).Milliseconds()))
+	dlt := time.Now()
+	blockDownloadMs.Observe(float64(dlt.Sub(start).Milliseconds()))
+	toVerify, err := blocks.NewROBlockSlice(results)
 	if err != nil {
+		log.WithError(err).WithFields(b.logFields()).Debug("Batch conversion to ROBlock failed")
+		return b.withRetryableError(err)
+	}
+
+	vb, err := w.cfg.verifier.verify(toVerify)
+	blockVerifyMs.Observe(float64(time.Since(dlt).Milliseconds()))
+	if err != nil {
+		if shouldDownscore(err) {
+			w.cfg.downscore(b.columns.peer, "invalid SignedBeaconBlock batch rpc response", err)
```

Comment:
This panicked for me, when b.columns is nil.

```suggestion
			w.cfg.downscore(b.peer, "invalid SignedBeaconBlock batch rpc response", err)
```